components/brave_wallet/browser/solana_instruction_builder.cc

```diff
@@ -0,0 +1,108 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/solana_instruction_builder.h"
+
+#include <type_traits>
+
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/browser/solana_account_meta.h"
+#include "brave/components/brave_wallet/browser/solana_instruction.h"
+#include "brave/components/brave_wallet/common/brave_wallet_types.h"
+#include "build/build_config.h"
+
+namespace {
+
+template <typename T>
+void UintToLEBytes(T val, std::vector<uint8_t>* bytes) {
+  static_assert(
+      std::is_same<T, uint64_t>::value || std::is_same<T, uint32_t>::value,
+      "Incorrect type passed to function UintToLEBytes.");
+
+  DCHECK(bytes);
+  size_t vec_size = sizeof(T) / sizeof(uint8_t);
+  *bytes = std::vector<uint8_t>(vec_size);
+
+  uint8_t* ptr = reinterpret_cast<uint8_t*>(&val);
+  for (size_t i = 0; i < vec_size; i++) {
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
+    bytes->at(i) = *ptr++;
+#else
+    bytes->at(vec_size - 1 - i) = *ptr++;
+#endif
+  }
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+namespace solana {
+
+namespace system_program {
+
+absl::optional<SolanaInstruction> Transfer(const std::string& from_pubkey,
+                                           const std::string& to_pubkey,
+                                           uint64_t lamport) {
+  if (from_pubkey.empty() || to_pubkey.empty())
+    return absl::nullopt;
+
+  // Instruction data is consisted of u32 instruction index and u64 lamport.
+  std::vector<uint8_t> instruction_data;
+  UintToLEBytes(static_cast<uint32_t>(SolanaSystemInstruction::kTransfer),
+                &instruction_data);
+
+  std::vector<uint8_t> lamport_bytes;
+  UintToLEBytes(lamport, &lamport_bytes);
+  instruction_data.insert(instruction_data.end(), lamport_bytes.begin(),
+                          lamport_bytes.end());
+
+  return SolanaInstruction(kSolanaSystemProgramId,
+                           std::vector<SolanaAccountMeta>(
+                               {SolanaAccountMeta(from_pubkey, true, true),
+                                SolanaAccountMeta(to_pubkey, false, true)}),
+                           instruction_data);
+}
+
+}  // namespace system_program
+
+namespace spl_token_program {
+
+absl::optional<SolanaInstruction> Transfer(
```

Comment:
See https://docs.rs/spl-token/latest/src/spl_token/instruction.rs.html#105 for the interface.